### PR TITLE
Add rubocop rule to enforce `Notes` in exploit module info

### PR DESCRIPTION
**Pull request by dwelch-r7**
Adds a rubocop rule to enforce the presence of a list of  `SideEffects` in the `Notes` section of the module info hash

https://github.com/rapid7/metasploit-framework/wiki/Definition-of-Module-Reliability,-Side-Effects,-and-Stability

# Verification
- [ ] Tests pass
- [ ] Nothing blows up when running `rubocop --only=Lint/SideEffectsInNotes modules/exploits`

### CODE REVIEW

File: lib/rubocop/cop/lint/side_effects_in_notes.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            add_offense(last_key || hash, message: NO_NOTES_MSG)
+          end
+        end
+
+        private
+
+        def check_for_side_effects(notes)
+          last_key = nil
+          side_effects_present = false
+          p notes
```
Comment (adfoster-r7) [inline lib/rubocop/cop/lint/side_effects_in_notes.rb:48]:
```suggestion
```

File: spec/rubocop/cop/lint/side_effects_in_notes_spec.rb [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+  it 'requires Notes to be present when no keys are present' do
+    expect_offense(<<~RUBY)
+      class DummyModule
+        def initialize(info = {})
+          super(
+            update_info(
+              info,
+              {}
+              ^^ Module is missing the Notes section [...]
```
Comment (adfoster-r7) [inline spec/rubocop/cop/lint/side_effects_in_notes_spec.rb:40]: We should probably link to the wiki page for more details to help the average user understand what this means
https://github.com/rapid7/metasploit-framework/wiki/Definition-of-Module-Reliability,-Side-Effects,-and-Stability

That aside, the wiki page also needs a bit of direction in it too 😄

File: .rubocop.yml [hunk 3]
```diff
@@ -158,6 +159,11 @@ Lint/ModuleDisclosureDatePresent:
     # Only exploits require disclosure dates, but they can be present in auxiliary modules etc.
     - 'modules/exploits/**/*'
 
+Lint/SideEffectsInNotes:
+  Include:
+    # Only exploits require disclosure dates, but they can be present in auxiliary modules etc.
+    - 'modules/exploits/**/*'
```
Comment (adfoster-r7) [inline .rubocop.yml:165]: I think this copy/pastas needs tweaked, notes should exist in auxiliary modules I think 👀

File: spec/rubocop/cop/lint/side_effects_in_notes_spec.rb (hunk 2 above)
Comment (h00die) [reply spec/rubocop/cop/lint/side_effects_in_notes_spec.rb:40]: An example on that page would be nice, I always have to look up old modules.
Prob a good idea to update the example modules as well

Comment (adfoster-r7) [reply spec/rubocop/cop/lint/side_effects_in_notes_spec.rb:40]: > you can also abbreviate offense messages with `[…]`
> https://www.rubydoc.info/gems/rubocop/RuboCop/RSpec/ExpectOffense

File: .rubocop.yml [hunk 4]
```diff
@@ -20,6 +20,7 @@ require:
   - ./lib/rubocop/cop/lint/module_disclosure_date_format.rb
   - ./lib/rubocop/cop/lint/module_disclosure_date_present.rb
   - ./lib/rubocop/cop/lint/deprecated_gem_version.rb
+  - ./lib/rubocop/cop/lint/side_effects_in_notes.rb
```
Comment (adfoster-r7) [inline .rubocop.yml:23]:
```suggestion
  - ./lib/rubocop/cop/lint/module_side_effects_in_notes.rb
```

File: lib/rubocop/cop/lint/side_effects_in_notes.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+module RuboCop
+  module Cop
+    module Lint
+      class SideEffectsInNotes < Base
+
+        NO_NOTES_MSG = 'Module is missing the Notes section which must include SideEffects - https://github.com/rapid7/metasploit-framework/wiki/Definition-of-Module-Reliability,-Side-Effects,-and-Stability'
+        NO_SIDE_EFFECTS_MSG = 'Module is missing SideEffects from the Notes section - https://github.com/rapid7/metasploit-framework/wiki/Definition-of-Module-Reliability,-Side-Effects,-and-Stability'
+
+
```
Comment (adfoster-r7) [inline lib/rubocop/cop/lint/side_effects_in_notes.rb:11]:
```suggestion
```

File: .rubocop.yml (hunk 3 above)
Comment (dwelch-r7) [reply .rubocop.yml:165]: Sure, I was wondering about this and I think it definitely makes sense for aux modules to have side effects, could easily imagine it leaving a trail in the logs

File: .rubocop.yml [hunk 6]
```diff
@@ -158,6 +159,12 @@ Lint/ModuleDisclosureDatePresent:
     # Only exploits require disclosure dates, but they can be present in auxiliary modules etc.
     - 'modules/exploits/**/*'
 
+Lint/SideEffectsInNotes:
```
Comment (adfoster-r7) [inline .rubocop.yml:162]:
```suggestion
Lint/ModuleSideEffectsInNotes:
```

File: .rubocop.yml [hunk 7]
```diff
@@ -158,6 +159,12 @@ Lint/ModuleDisclosureDatePresent:
...
+  Include:
+    # Only exploits and auxiliary modules require SideEffects to be listed.
+    - 'modules/exploits/**/*'
+    - 'modules/auxiliary/**/*'
```
Comment (jmartin-r7) [inline .rubocop.yml:166]: Let's add `modules/post` as well.

And per @adfoster-r7 's idea we could expand this to include `Stability` and `Reliability` keys.

File: spec/rubocop/cop/lint/side_effects_in_notes_spec.rb (hunk 2 above)
Comment (wvu-r7) [reply spec/rubocop/cop/lint/side_effects_in_notes_spec.rb]: > Note that I did reorder module traits to stability, reliability, and then side effects due to the following thought process:
>
> 1. Will I crash anything if I run this?
> 2. Can I expect to get a shell?
> 3. What might need to be cleaned up?

#11785

**Comment by adfoster-r7** [comment]
## Release Notes

Adds automated tooling to ensure that all modules now contain additional metadata on their Reliability, Side Effects, and Stability

**Pull request merged**